### PR TITLE
Story/s3 repository can be configured to authenticate using aws ec2 instance metadata

### PR DESCRIPTION
**Pull request by kiddouk**
Any of the checked boxes below indicate that I took action:
- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [X] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [X] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:
- [X] Before submitting a pull request, I started a discussion on the
  [Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
  or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [X] I wrote a design document. A design document can be
  brief but explains the use case or problem you are trying to solve,
  touches on the planned implementation approach as well as the test cases
  that verify the behavior. Optimally, design documents should be submitted
  as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
  can be found in the Gradle GitHub repository. Please let us know if you need help with
  creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
  test coverage to verify the behavior. Before submitting the pull request
  I ran a build on your local machine via the command
  `./gradlew quickCheck <impacted-subproject>:check`.
- [x] The pull request updates the Gradle documentation like user guide,
  DSL reference and Javadocs where applicable.

===== SUMMARY =====
- This is related to
  https://github.com/gradle/gradle/blob/master/design-docs/finding-and-using-credentials.md#story-an-s3-repository-can-be-configured-to-authenticate-using-awss-ec2-instance-metadata
- Note that the AWS S3 Client implementation will now use the following
  credentials (in this specific order)
  - Environment (AWS_ACCESS_KEY, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN)
  - Java System Properties - aws.accessKeyId and aws.secretKey
  - Credential profiles file at the default
    location (~/.aws/credentials) shared by all AWS SDKs and the AWS CLI
  - Instance Profile Credentials - delivered through the Amazon EC2 metadata service
- The implementation assumes that only ONE authentication can be used
  per s3-resource at a time
- This patch enforces that each Authentication now declares
  `requiresCredentials` to be explicit
- Exposes `sessionToken` in the DSL for `AwsCredentials`

### CODE REVIEW

**Comment by kiddouk** [comment]
I am missing some unit tests here but I hope that @wolfs can tell me if this PR is going in the right direction or not.

Also, looking at the story, it is advised to implement a HTTP Stub of AWS Metadata fetch. Can the reviewer point me in the right direction to achieve this ?


Cheers,

**Comment by kiddouk** [comment]
I ended up finishing it all up since I was in the zone. The only thing that I am missing is spying on Environment loading but it is not as easy as one think since the Role (and/or var env) are not loaded when we create the client but when an API call is emitted.

**Comment by kiddouk** [comment]
Thinking a bit about that PR, I should probably expose the sessionToken in the `AwsIamCredentials` DSL. This way, gradle would now support KEY/SECRET/TOKEN from EC2 Roles, Environment Vars (both are already the case here) and in the DSL.

For info, the session token is used for EC2 Roles (IAM), 2FA authentication if your KEY/SECRET are bound to a 2FA device, and in credentials providers (example : https://github.com/hashicorp/vault).

I also think that this fits in the scope of this story and makes it more complete in some way.

File: subprojects/resources-http/src/main/java/org/gradle/authentication/http/BasicAuthentication.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 /**
  * Authentication scheme for basic access authentication over HTTP. When using this scheme, credentials are sent preemptively.
  */
+
```
Comment (wolfs) [inline subprojects/resources-http/src/main/java/org/gradle/authentication/http/BasicAuthentication.java]: Remove this newline

**Comment by wolfs** [comment]
I am still looking through your changes. One thing to note is that two integration tests are failing: `MavenPublishS3IntegrationTest` and `MavenS3RepoErrorsIntegrationTest`.
You can execute all integration tests for `resourcesS3` by running
```
./gradle resourcesS3:integTest
```

**Comment by kiddouk** [comment]
Thanks for looking into this. I will have a look on my side and fix those.

**Comment by wolfs** [comment]
Adding the import `import org.gradle.authentication.aws.AwsIamAuthentication` to the build script for `org.gradle.integtests.resource.s3.maven.MavenPublishS3IntegrationTest#can publish to a S3 Maven repository with IAM` makes it fail at the right point: 
```
Could not write to resource 's3://tests3Bucket/maven/org/gradle/test/publishS3Test/1.0/publishS3Test-1.0.jar'.
      > Unable to load AWS credentials from any provider in the chain
```

That is probably where `Add integration test support to stub out AWS IM http api calls.` from the design doc comes into play.

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ConnectorFactory.java [hunk 2]
```diff
@@ -64,6 +64,6 @@ public ExternalResourceConnector createResourceConnector(ResourceConnectorSpecif
             }
         }
 
-        throw new IllegalStateException("S3 resource should either specify AwsIamAutentication or provide some AwsCredentials.");
```
Comment (wolfs) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ConnectorFactory.java]: Why did you change that back to the old exception? Changing the test would probably have been better.

**Comment by kiddouk** [comment]
I have now fixed the unit tests.

A note on the fix:
Since the credentials and fetched in the *client*, it is impossible to mock an IAM Server (like S3Server is currently mocked). Indeed, the client is trying to reach `169.254.169.254` for the instance meta data.

Instead, since we use the whole provider chain in AWS, we let the client fallback to the `properties` as explained [in the aws java sdk documentation](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html)

This still ensures that the client is not bound to `BasicAwsCredentials`.


Other possible scenario , we could use [JUnit system Rules](http://stefanbirkner.github.io/system-rules/) to inject AWS_ACCESS_SECRET_KEY/AWS_ACCESS_KEY_ID but this would add a dependency for testing. Ultimately, we need to trust that the `DefaultCredentialsChainProvider` is properly tested in the aws sdk code base.

Let me know your thoughts on this. I'd be happy to revisit if needed.

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ConnectorFactory.java (hunk 2 above)
Comment (kiddouk) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ConnectorFactory.java]: Ok, noted. Will revert and change the test.

File: subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.gradle.authentication.aws;
+
+import org.gradle.api.Incubating;
+import org.gradle.authentication.Authentication;
+
+/**
+ * Authentication scheme for digest access authentication over HTTP.
+ */
```
Comment (bmuschko) [inline subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java]: Missing `@since` annotation.

File: subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/package-info.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Classes related to transport authentication protocols for HTTP.
+ *
+ * @since 2.7
```
Comment (bmuschko) [inline subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/package-info.java]: The version should be 3.1.

File: subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+@Incubating
+public interface AwsIamAuthentication extends Authentication {
```
Comment (bmuschko) [inline subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java]: There's a typo in the class name. Should be `AwsImAuthentication` - `Im` as in "instance metadata". All other code would have to be aligned to follow the class name pattern.

File: subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/package-info.java (hunk 4 above)
Comment (kiddouk) [inline subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/package-info.java]: roger that.

File: subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/maven/MavenPublishS3IntegrationTest.groovy [hunk 6]
```diff
@@ -77,4 +77,56 @@ publishing {
         module.assertPublishedAsJavaModule()
         module.parsedPom.scopes.isEmpty()
     }
+
+  
+    def "can publish to a S3 Maven repository with IAM"() {
+        given:
+        System.setProperty('aws.accessKeyId', 'someKey')
```
Comment (bmuschko) [inline subprojects/resources-s3/src/integTest/groovy/org/gradle/integtests/resource/s3/maven/MavenPublishS3IntegrationTest.groovy]: I'd think that the system properties would have to be provided to the `executor` instance.

File: subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java (hunk 5 above)
Comment (kiddouk) [inline subprojects/resources-s3/src/main/java/org/gradle/authentication/aws/AwsIamAuthentication.java]: Ah ! Funny enough, I thought that "Im" was a typo for "Iam" :) Fair enough.

**Comment by bmuschko** [comment]
@Kallin Can you also add `@Incubating` and `@since` to all new public API methods or classes that were introduced with this PR?

**Comment by kiddouk** [comment]
Since @Kallin hasn't participated to gradle since 2014 (and was an external contributor back then), I guess that you have made a typo in the name and meant @kiddouk instead ? :)

**Comment by kiddouk** [comment]
Not sure about why this PR is still tagged as `waiting-for-contributor` as I think that I have implemented everything that was requested.

Let me know otherwise, I'd be happy to contribute more on this.

**Comment by devinlundberg** [comment]
Anything I can do to help this move forward?

**Comment by kiddouk** [comment]
@devinlundberg it is a bit unclear to me as wether gradle team is waiting for "something" or this PR just sits until it gets a chance to be merged as they might be busy on other things.

**Comment by lacasseio** [comment]
@devinlundberg I'm picking up this PR (squashing/rebasing) in order to ensure nothing is failing. I will keep up you posted on the process.

**Comment by kiddouk** [comment]
ping. :)

**Comment by lacasseio** [comment]
We are looking at merging this really soon. So far, everything looks good. We need to do a final check and it should be good to go. Thank you for your patience.

**Comment by lacasseio** [comment]
It was longer than expected, however, we made it. This PR was manually squashed, rebase and merge in [this commit](https://github.com/gradle/gradle/commit/7283def6e4c9773ed92340e57ae5c2033b783c05). Huge thanks to @kiddouk for his collaboration and patience on this PR.

**Comment by kiddouk** [comment]
Thanks for "merging" it @lacasseio
